sympy/utilities/decorator.py

```diff
@@ -193,37 +195,13 @@ def public(obj):
     return obj
 
 
-class ClassPropertyDescriptor(object):
-    """
-    Helper class for ``classproperty`` decorator.
-    """
-    def __init__(self, fget):
-        self.fget = fget
-
-    def __get__(self, obj, klass=None):
-        if klass is None:
-            klass = type(obj)
-
-        return self.fget.__get__(obj, klass)()
-
-
-def classproperty(func):
-    """
-    This decorator can be used to define a property of a class
-    which is intended to be accessed using class and objects both.
-
-    Examples
-    ========
-
-    >>> from sympy.utilities.decorator import classproperty
-    >>> class Q(object):
-    ...     @classproperty
-    ...     def a_property(self):
-    ...         return 2
-    >>> Q.a_property
-    2
-    """
-    if not isinstance(func, (classmethod, staticmethod)):
-        func = classmethod(func)
-
-    return ClassPropertyDescriptor(func)
+def memoize_property(storage):
+    """Create a property, where the lookup is stored in ``storage``"""
```

Comment:
The docstring is indeed less explanatory, I'll fix that up.

This decorator actually makes sure that the returned value has only one instance.